### PR TITLE
Loader

### PR DESCRIPTION
**Pull request by BenGriffith**

### CODE REVIEW

File: tests/unit/test_load.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import os
+from datetime import datetime
+
+from google.cloud.storage.bucket import Bucket
+
+
+def test_load_get_bucket(load):
+    assert type(load._get_bucket()) == Bucket
```
Comment (bbelderbos) [inline tests/unit/test_load.py:8]: not sure if we would use `is` here?

File: tests/unit/test_load.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    assert load._get_bucket().name == os.getenv("BUCKET")
```
Comment (bbelderbos) [inline tests/unit/test_load.py:9]: should we mock os? I mean is it reliable?

File: tests/conftest.py [hunk 3]
```diff
@@ -15,4 +20,21 @@ def coinmarket_invalid_response():
     connection = Connection()
     connection._headers["X-CMC_PRO_API_KEY"] = ""
     response = connection.request(url=LISTINGS_LATEST_URL)
-    return response
\ No newline at end of file
+    return response
+
+
+@pytest.fixture
```
Comment (bbelderbos) [inline tests/conftest.py:26]: nice use of fixtures (and putting them here)

File: src/crypto/utils/api.py [hunk 4]
```diff
@@ -1,18 +1,19 @@
...
+
 import requests
-from decouple import config
 
 
 class Connection:
 
-    def __init__(self) -> None:
-        self._api_key = config("COINMARKET_API_KEY", cast=str)
+    def __init__(self, limit: str = 5000) -> None:
```
Comment (bbelderbos) [inline src/crypto/utils/api.py:8]: 5000 != str, type hint as int?

File: src/crypto/transform.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    def _load_facts(self) -> None:
+        pass
+
+    def _load_dimensions(self) -> None:
+        pass
+
+
+if __name__ == "__main__":
+    transform = Transform()
+    transform._get_blob()
```
Comment (bbelderbos) [inline src/crypto/transform.py:39]: internal methods "should" not be called, make "public"? so no underscore? 
https://youtu.be/ZPdpmiWdDm8?t=22

File: src/crypto/transform.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+from google.cloud.storage.bucket import Bucket
+from google.cloud.storage.blob import Blob
+
+
+class Transform:
+
+    def __init__(self) -> None:
+        self._blob_client = BlobClient()
+        self._bq_client = BigQueryClient()
+        self._blob = f"{datetime.today().strftime('%Y-%m-%d')}.json"
```
Comment (bbelderbos) [inline src/crypto/transform.py:16]: saw this in several places, maybe abstract into helper?

File: .gitignore [hunk 7]
```diff
@@ -127,6 +127,7 @@ venv/
 ENV/
 env.bak/
 venv.bak/
+service_account/
```
Comment (bbelderbos) [inline .gitignore:130]: what is here?

File: .env-template [hunk 8]
```diff
@@ -1 +1,3 @@
 COINMARKET_API_KEY=
+BUCKET=
```
Comment (bbelderbos) [inline .env-template:2]: template file, good convention

File: src/crypto/load.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import json
+from typing import Dict
+from datetime import datetime
+
+from google.cloud.storage import Client
+from google.cloud.storage.bucket import Bucket
+from google.oauth2 import service_account
+
+from crypto.utils.api import Connection
+from crypto.utils.constants import LISTINGS_LATEST_URL
```
Comment (bbelderbos) [inline src/crypto/load.py:11]: nice pep8 separated imports

File: src/crypto/load.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class Load:
+
+    def __init__(self) -> None:
+        self._service_account = os.getenv("CLOUD_STORAGE")
```
Comment (bbelderbos) [inline src/crypto/load.py:17]: could this be constructor arg to "inject" it? so we are not dependent on os environ only? might be handy when testing

File: src/crypto/load.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        self._credentials = service_account.Credentials.from_service_account_file(self._service_account)
```
Comment (bbelderbos) [inline src/crypto/load.py:18]: in that sense maybe make it an option to inject credentials or even the whole client instance?

File: src/crypto/load.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        self._gcs_client = Client(credentials=self._credentials)
+    
+    def _get_bucket(self) -> Bucket:
+        bucket_name = os.getenv("BUCKET")
+        bucket = self._gcs_client.bucket(bucket_name)
+        return bucket
+
+    def create_blob(self, name: str, crypto_data: Dict) -> None:
```
Comment (bbelderbos) [inline src/crypto/load.py:26]: >= 3.9 you can use `dict` for typing

File: src/crypto/load.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        bucket = self._get_bucket()
+        blob_name = name
```
Comment (bbelderbos) [inline src/crypto/load.py:28]: seems redundant

File: tests/conftest.py [hunk 14]
```diff
@@ -15,4 +20,21 @@ def coinmarket_invalid_response():
...
+def load():
+    return Load()
+
+
+@pytest.fixture
+def gcs_client():
+    return Load()._gcs_client
```
Comment (bbelderbos) [inline tests/conftest.py:33]: maybe not "internal" then?

File: tests/unit/test_load.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def test_load_create_blob(coinmarket_response, load, bucket):
+    blob_name = f"{datetime.today().strftime('%Y-%m-%d')}-unit-test.json"
+    load.create_blob(name=blob_name, crypto_data=coinmarket_response)
+    blob = bucket.blob(blob_name)
+    blob_exists = blob.exists()
+    assert blob_exists
+    if blob_exists:
```
Comment (bbelderbos) [inline tests/unit/test_load.py:18]: teardown code

File: tests/unit/test_load.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+    assert type(load._get_bucket()) == Bucket
+    assert load._get_bucket().name == os.getenv("BUCKET")
+
+
+def test_load_create_blob(coinmarket_response, load, bucket):
+    blob_name = f"{datetime.today().strftime('%Y-%m-%d')}-unit-test.json"
+    load.create_blob(name=blob_name, crypto_data=coinmarket_response)
+    blob = bucket.blob(blob_name)
+    blob_exists = blob.exists()
+    assert blob_exists
```
Comment (bbelderbos) [inline tests/unit/test_load.py:17]: ah I see why the extra var, to not call `blob.exists()` twice right?

**Review by bbelderbos — COMMENTED** [review]
Great work!

File: tests/unit/test_load.py (hunk 1 above)
Comment (BenGriffith) [reply tests/unit/test_load.py:8]: yes, I'll update it

File: tests/unit/test_load.py (hunk 2 above)
Comment (BenGriffith) [reply tests/unit/test_load.py:9]: I made some updates to the use of `os` that will come through in next couple PRs. Also, I want to say it should be reliable cause will be inside of container and env variables passed in through `docker-compose`. I haven't implemented it yet but I think it'll make more sense once everything is in place.

File: src/crypto/utils/api.py (hunk 4 above)
Comment (BenGriffith) [reply src/crypto/utils/api.py:8]: good catch

File: src/crypto/transform.py (hunk 5 above)
Comment (BenGriffith) [reply src/crypto/transform.py:39]: yes, good call. I'll rework this.

File: src/crypto/transform.py (hunk 6 above)
Comment (BenGriffith) [reply src/crypto/transform.py:16]: great suggestion. I started working on making some helper(s) last night to address this. Thanks.

File: .gitignore (hunk 7 above)
Comment (BenGriffith) [reply .gitignore:130]: planning to store the service account json files here and then mount this directory to containers and then use the os.getenv to get service account details and pass to GCP services for auth

File: src/crypto/load.py (hunk 10 above)
Comment (BenGriffith) [reply src/crypto/load.py:17]: good call. I'll work on this.

File: src/crypto/load.py (hunk 11 above)
Comment (BenGriffith) [reply src/crypto/load.py:18]: yeah, I think might go the route of the whole client

File: src/crypto/load.py (hunk 12 above)
Comment (BenGriffith) [reply src/crypto/load.py:26]: wasn't aware of that, thank you

File: tests/conftest.py (hunk 14 above)
Comment (BenGriffith) [reply tests/conftest.py:33]: you're right. I'll rework this. Thanks.

File: tests/unit/test_load.py (hunk 16 above)
Comment (BenGriffith) [reply tests/unit/test_load.py:17]: yeah, it was kinda funky. Was late when I wrote this so not the greatest implementation and mainly wanted to get it working before signing off. I'm going to rework the test/approach.

File: .gitignore (hunk 7 above)
Comment (BenGriffith) [reply .gitignore:130]: just occurred to me, I might need to have a `service_account_template` directory similar to `.env` and `.env-template`

File: tests/unit/test_load.py (hunk 2 above)
Comment (bbelderbos) [reply tests/unit/test_load.py:9]: thanks, good point

File: .gitignore (hunk 7 above)
Comment (bbelderbos) [reply .gitignore:130]: good idea

File: tests/unit/test_load.py (hunk 16 above)
Comment (bbelderbos) [reply tests/unit/test_load.py:17]: ok cool, let me know if you get stuck on anything ...

**Pull request merged**